Feature: Add diago_dav_subspace module to pyabacus

a1henu opened the pull request:
This commit adds the `diago_dav_subspace` module to the `pyabacus` package, which is responsible for diagonalizing the Hamiltonian matrix using Davison subspace method.

### Reminder
- [x] Have you linked an issue with this pull request?

### Linked Issue
#4882

a1henu committed:
<Feature>[pyabacus]: Add diago_dav_subspace module to pyabacus

This commit adds the `diago_dav_subspace` module to the `pyabacus` package, which is responsible for diagonalizing the Hamiltonian matrix using Davison subspace method.

The current Pythonization version targets matrix diagonalization. The original algorithm does not require passing a matrix; instead, it uses a function pointer to pass a linear operator for diagonalization.

So the current implementation still needs to wrap std::function<void(T*, T*, const int, const int, const int, const int)> to a Python callable type, which requires further iteration.

a1henu committed:
Merge branch 'develop' into devdev

a1henu committed:
Add Python function and class signatures for ModuleBase and hsolver to improve readability and enhance type checking in Python

a1henu committed:
Merge branch 'devdev' of https://github.com/a1henu/abacus-develop into devdev

jieli-matrix commented on python/pyabacus/src/py_diago_dav_subspace.cpp:
```diff
@@ -0,0 +1,88 @@
+{
+    py::module hsolver = m.def_submodule("hsolver");
+
+    py::class_<hsolver::diag_comm_info>(hsolver, "diag_comm_info")
+        .def(py::init<const int, const int>(), "rank"_a, "nproc"_a)
+        .def_readonly("rank", &hsolver::diag_comm_info::rank)
+        .def_readonly("nproc", &hsolver::diag_comm_info::nproc);
```
You add `rank` and `nproc` here and I think it corresponds to the MPI Rank. However, you need to call `mpi4py` to run MPI in pyabacus. See the [playground](https://github.com/jieli-matrix/pybind-mpi/tree/main) and try!

a1henu replied:
I plan to add parallel processing capabilities in the next PR. I will review the playground for reference. Thank you!

jieli-matrix commented on python/pyabacus/src/py_diago_dav_subspace.cpp:
```diff
@@ -0,0 +1,88 @@
+        .def_readonly("rank", &hsolver::diag_comm_info::rank)
+        .def_readonly("nproc", &hsolver::diag_comm_info::nproc);
+
+    py::class_<hsolver::Diago_DavSubspace<std::complex<double>, base_device::DEVICE_CPU>>(hsolver, "Diago_DavSubspace")
+        .def(py::init<const std::vector<double>, const int, const int,
+                    const int, const double, const int, const bool,
+                    const hsolver::diag_comm_info>(),
+        "precondition"_a, "nband"_a, "nbasis"_a, "david_ndim"_a, "diag_thr"_a, "diag_nmax"_a, "need_subspace"_a, "diag_comm"_a)
+        .def("diag",
+            [] (hsolver::Diago_DavSubspace<std::complex<double>, base_device::DEVICE_CPU>& self,
+                py::array_t<std::complex<double>> h_mat,
+                py::array_t<std::complex<double>> psi_in,
+                const int psi_in_dmax,
+                py::array_t<double> eigenvalue_in,
+                const std::vector<bool>& is_occupied,
+                const bool& scf_type)
```
Good. However, Davidson Method should be designed as a pure math class(function). It means you can call the ABACUS API in your inner codes but expose the python function as a pure math class. Plz see the [Davidson implementation](https://github.com/pyscf/pyscf/blob/master/pyscf/lib/linalg_helper.py#L200-L225) in PySCF.

jieli-matrix replied:
Also, plz add `rdoc` to show the docs in pyabacus.

a1henu replied:
i have added `pbdoc` in pyabacus and designed new interface

jieli-matrix commented:
A good practice -- the first contribution to pyabacus from the external users！🎉
However, I think the try will start from the one-one mapping (ABACUS <-> pyabacus) but not stop at such level: math operations will separate from the physics features, and it would be the cornerstone of numerical simulations.

I list the TODOs as follows:

- [ ] Add the tests with comparison to `scipy.linalg.eigh`, Refer to tests [here](https://github.com/pyscf/pyscf/blob/master/pyscf/lib/test/test_linalg_helper.py#L55-L94). Both precision and performance.
- [ ] Design the python APIs so that it can be treated as a pure math feature.
- [ ] Plz add docs in pyabacus like [this](https://github.com/jieli-matrix/abacus-develop/blob/develop/python/pyabacus/src/py_m_nao.cpp#L20-L28), to notify users which case your Davidson methods can work(Hermitian matrix or non-sym matrix)? 
- [ ] (optional): with the tests results from the first item, you may need to add MPI support to pyabacus. I have set up the environment for you. Try it!

jieli-matrix commented on python/pyabacus/tests/test_diago_dav_subspace.py:
```diff
@@ -0,0 +1 @@
+from __future__ import annotations
```
Plz add the tests compared with `scipy`!

a1henu replied:
i have added tests :)

jieli-matrix commented on python/pyabacus/src/pyabacus/ModuleBase/_module_base.py:
```diff
@@ -0,0 +1,105 @@
+
+import numpy as np
+
+from typing import overload
+from numpy.typing import NDArray
+
+class Sphbes:
+    def __init__(self) -> None: ...
+    @overload
+    @staticmethod
+    def sphbesj(l: int, x: float) -> float: ...
+    @overload
+    @staticmethod
+    def sphbesj(
+        n: int, 
+        r: NDArray[np.float64], 
+        q: int, 
+        l: int, 
+        jl: NDArray[np.float64]
+    ) -> None: ...
+    @overload
+    @staticmethod
+    def dsphbesj(l: int, x: float) -> float: ...
+    @overload
+    @staticmethod
+    def dsphbesj(
+        n: int, 
+        r: NDArray[np.float64], 
```
I was wondering whether it was necessary to add function signatures in python's `__init__.py`. Plz refer to this repo's [example](https://gitlab.com/veloxchem/veloxchem/-/blob/master/src/pymodule/__init__.py?ref_type=heads), maybe the suitable code looks like
` from ._core import ModuleBase`

a1henu replied:
hmm...Using only Pybind, we cannot enable IDEs to provide type hints and code completion. Adding function signatures in Python files can achieve this.

jieli-matrix replied:
I see. You can keep them in your local changes (since they only provide convenience for coding not for usage), but keep `__init__` with only necessary information like the example I show you.

a1henu committed:
Merge remote-tracking branch 'upstream/develop' into devdev

a1henu committed:
modify dav_subspace, now the module could work but the result is wrong

a1henu committed:
Fix memory issue in Diago_DavSubspace destructor

The destructor of Diago_DavSubspace was causing a memory issue, preventing the program from running correctly. Lines 72-74 were previously commented out to avoid the issue, but this led to a memory leak. This commit addresses the root cause of the memory issue and ensures proper memory deallocation without causing a crash.

- Uncommented lines 72-74 in the destructor.

This ensures that all allocated memory is properly deallocated, preventing memory leaks but still unable to run the `pyabacus::hsolver::dav_subspace()`

a1henu committed:
add pbdoc to the pybind file and add document to the python func signatures

a1henu committed:
modify some docs

a1henu committed:
Merge remote-tracking branch 'upstream/develop' into devdev

a1henu committed:
fixed the memory issue in dav_subspace, add an example(python/pyabacus/diago_matrix.py)

a1henu committed:
add pytest to diago_dav_subspace

a1henu committed:
delete an matrix

a1henu committed:
modify _hsolver.py

a1henu committed:
Refactor diagonalization to accept operators instead of explicit matrices

Previously, our module could only accept explicit matrices for diagonalization. I have now modified the code to accept operators as parameters for diagonalization, eliminating the need to store the entire matrix explicitly. This allows us to compute eigenvalues for larger Hamiltonians using the sparse matrix storage provided by `SciPy`.

The operator is a function pointer that accepts a vector and returns a vector. To diagonalize an operator A, we define a function or lambda function `mv_op` such that `mv_op(x) = Ax`, which can then be used in the diagonalization process.

Additionally, I have added a new test case for the Hamiltonian matrix corresponding to H2O. The matrix size is `67024x67024`, which previously could not be stored explicitly due to memory constraints. The updated operator method now supports the computation of eigenvalues for such large matrices.

Test cases have been updated to reflect these changes.

a1henu committed:
Refactor: Replace manual loop with std::copy for better readability and potential performance improvement

- Replaced manual loop with `std::copy` to copy elements from `psi_in` to `psi` and from `hpsi_ptr` to `hpsi_out`.
- This change simplifies the code, enhances readability and may improve performance due to optimized standard library `std::copy` implementations.

a1henu committed:
Refactor hpsi_func to support matrix-matrix multiplication operator and rename two variables' name

- Modified `hpsi_func` to handle matrix-matrix multiplication instead of vector operations.
- This vectorization significantly improved computation speed.
- Renamed variables in the Python frontend interface:
  - `nbasis` to `dim`
  - `nband` to `num_eigs`
- These changes align with the module's design as a pure mathematical interface.

a1henu committed:
Merge branch 'develop' into devdev

jieli-matrix commented on python/pyabacus/tests/test_diago_dav_subspace.py:
```diff
@@ -0,0 +1,48 @@
+    diag_elem = np.where(np.abs(diag_elem) < 1e-5, 1e-5, diag_elem)
+    precond = 1.0 / np.abs(diag_elem)
+
+    e, _ = hsolver.dav_subspace(
+        mm_op,
+        v0,
+        nbasis,
+        nband,
+        precond,
+        dav_ndim=8,
+        tol=1e-4,
+        max_iter=5000,
+        scf_type=False
+    )
+    
+    return e
+
+def diag_eigsh(h_sparse, nband):
+    e, _ = scipy.sparse.linalg.eigsh(h_sparse, k=nband, which='SA', maxiter=5000, tol=1e-4)
+    return e
```
It seems you set the stopping criterion to `1e-4`. I don't think it is a good stopping criterion since the threshold of SCF is at least 1e-7. If possible, try the `1e-7` and machine precision.

a1henu replied:
i have changed it.

jieli-matrix commented on python/pyabacus/tests/test_diago_dav_subspace.py:
```diff
@@ -0,0 +1,48 @@
+    h_sparse = scipy.io.loadmat(file_name)['Problem']['A'][0, 0]
+    e_pyabacus = diag_pyabacus(h_sparse, nband)
+    e_scipy = diag_eigsh(h_sparse, nband)
+    np.testing.assert_allclose(e_pyabacus, e_scipy, atol=1e-2)
```
oh, `1e-2` is so coarse for numerical precision. Plz set `atol` to `1e-7` and test the relative errors.

a1henu replied:
I've changed it~

a1henu committed:
Enhance test cases to improve precision requirements

- Updated test cases to ensure absolute error is less than 1e-8.
- Added a new test case for random sparse matrices to validate the implementation.

a1henu committed:
update README.md

a1henu committed:
update `README.md` and `pyproject.toml`

a1henu committed:
Merge branch 'develop' into devdev

mohanchen commented:
It seems that the PR has sufficient content to be merged. Further updates are welcome.
